File: spec/queue_spec.rb
# frozen_string_literal: true

require_relative '../lib/queue.rb'

RSpec.describe 'Queue - Ruby' do
  queue = Fifo.new

  it 'has an `empty?` method' do
    expect(queue).to respond_to(:empty?)
  end

  it 'is empty when newly created' do
    expect(queue.empty?).to be true
  end

  it 'has a `enqueue` method' do
    expect(queue).to respond_to(:enqueue)
  end

  it 'has a `dequeue` method' do
    expect(queue).to respond_to(:dequeue)
  end

  it '`dequeue` method returns the first enqueued value' do
    queue = Fifo.new

    queue.enqueue(40)
    queue.enqueue(50)
    expect(queue.dequeue).to eq(40)
  end

  it '`dequeue` method removes the first value from the store' do
    queue = Fifo.new

    queue.enqueue(40)
    queue.enqueue(50)
    queue.dequeue
    expect(queue.dequeue).to eq(50)
  end
end

File: spec/queue.spec.js
'use strict'

// Allow chai syntax like `expect(foo).to.be.ok;`
// jshint -W030

const chai = require('chai')
const chaiAsPromised = require('chai-as-promised')

chai.use(chaiAsPromised)

const expect = chai.expect

const Queue = require('../lib/queue')

describe('Queue - JS', function () {
  let queue = new Queue()

  it('has an `isEmpty` method', function () {
    expect(queue).to.respondTo('isEmpty')
  })

  it('is empty when newly created', function () {
    expect(queue.isEmpty()).to.be.true
  })

  it('has a `enqueue` method', function () {
    expect(queue).to.respondTo('enqueue')
  })

  it('has a `dequeue` method', function () {
    expect(queue).to.respondTo('dequeue')
  })

  it('`dequeue` method returns the first enqueued value', function () {
    queue = new Queue()

    queue.enqueue(100)
    queue.enqueue(200)
    expect(queue.dequeue()).to.eq(100)
  })

  it('`dequeue` method removes the first value from the store', function () {
    queue = new Queue()

    queue.enqueue(100)
    queue.enqueue(200)
    queue.dequeue()
    expect(queue.dequeue()).to.eq(200)
    expect(queue.dequeue()).to.be.undefined
  })
})
